### PR TITLE
WIP: Fix for Teams in HIP-Clang

### PR DESCRIPTION
**Pull request by arghdos**
This PR does the following:

- Adds a few missing HIPHostPinnedSpace methods that LAMMPS complains about (largely copy-pasted from the CUDA equivalents)  
- Implements a similar algorithm for `hip_get_max_block_size` and `hip_get_opt_block_size`, with some workarounds for bugs in 3.5 (notably in the `maxThreadsPerBlock` attribute), and some (currently) hardcoded values for the max waves / CU (to match the [Occupancy API's, which are more correct](https://github.com/ROCm-Developer-Tools/HIP/blob/a0b5dfd625d99af7e288629747b40dd057183173/vdi/hip_platform.cpp#L741)).
- Defaults all kernels w/o a launch bounds attributes to `__launch_bounds__(1024, 1)` to ensure we always generate correct code.
- Places `abort` behind a no-inline wrapper such that we minimize the impact of the generated printf code (which is rather hefty, and will inline everywhere by default).  Also, we manually implement the NDEBUG behaviour, as `assert` is ... spotty... in respecting this in 3.5

This fixes the triply nested parallelism test, which has been re-enabled.  I have one test that appears to be failing on my system (`HIP.IncrTest_12a_ThreadScratch`).  I believe this is related to the `1024` block-size assumption I made in `hip_internal_get_block_size`, but need to dig further.

### CODE REVIEW

File: core/src/HIP/Kokkos_HIP_Abort.hpp [hunk 1]
```diff
@@ -53,16 +53,26 @@
...
+#endif
+  abort();
 }
 
 }  // namespace Impl
 }  // namespace Kokkos
 
+#define HIP_ABORT(msg)                                                \
+  do {                                                                \
+    Kokkos::Impl::__hip_abort(msg, __FILE__, __LINE__, __FUNCTION__); \
```
Comment (Rombur) [inline core/src/HIP/Kokkos_HIP_Abort.hpp:74]: Let's not print FILE, LINE, and FUNCTION. `hip_abort` is only called inside `abort()` so we won't get any useful information and that way we can use a function instead of a macro.

File: core/src/HIP/Kokkos_HIP_Parallel_Range.hpp [hunk 2]
```diff
@@ -555,10 +555,8 @@ class ParallelScanHIPBase {
       // correctly, the unit tests fail with wrong results
       const int gridMaxComputeCapability_2x = 0x01fff;
 
-      // FIXME_HIP block sizes greater than 256 don't work correctly,
-      // the unit tests fail with wrong results
       const int block_size =
-          std::min(static_cast<int>(local_block_size(m_functor)), 256);
+          std::min(static_cast<int>(local_block_size(m_functor)), 1024);
```
Comment (Rombur) [inline core/src/HIP/Kokkos_HIP_Parallel_Range.hpp:559]: Just remove the `std::min`. That was just a temporary hack.

File: core/src/HIP/Kokkos_HIP_Parallel_Team.hpp [hunk 3]
```diff
@@ -165,8 +160,7 @@ class TeamPolicyInternal<Kokkos::Experimental::HIP, Properties...>
         static_cast<size_t>(vector_length()),
         static_cast<size_t>(team_scratch_size(0)) + 2 * sizeof(double),
         static_cast<size_t>(thread_scratch_size(0)) + sizeof(double));
-    // FIXME_HIP See comment about teams larger than 256 above
-    return std::min(block_size / vector_length(), 256);
+    return std::min(block_size / vector_length(), 1024);
```
Comment (Rombur) [inline core/src/HIP/Kokkos_HIP_Parallel_Team.hpp:163]: Remove the `std::min`

File: core/src/HIP/Kokkos_HIP_Parallel_Team.hpp [hunk 4]
```diff
@@ -701,9 +689,7 @@ class ParallelReduce<FunctorType, Kokkos::TeamPolicy<Properties...>,
     int64_t threadid = 0;
     if (m_scratch_size[1] > 0) {
       __shared__ int64_t base_thread_id;
-      // FIXME_HIP This uses g_device_hip_lock_arrays which is not working
       if (hipThreadIdx_x == 0 && hipThreadIdx_y == 0) {
-        Impl::hip_abort("Error should not be here (not implemented yet)\n");
```
Comment (Rombur) [inline core/src/HIP/Kokkos_HIP_Parallel_Team.hpp:706]: Why did you remove this code? `g_device_hip_lock_arrays` is not implemented in this PR.

File: core/src/HIP/Kokkos_HIP_Space.cpp [hunk 5]
```diff
@@ -445,6 +445,46 @@ void* SharedAllocationRecord<Kokkos::Experimental::HIPSpace, void>::
   return r_new->data();
 }
 
+void* SharedAllocationRecord<Kokkos::Experimental::HIPHostPinnedSpace, void>::
+    allocate_tracked(const Kokkos::Experimental::HIPHostPinnedSpace& arg_space,
+                     const std::string& arg_alloc_label,
+                     const size_t arg_alloc_size) {
+  if (!arg_alloc_size) return (void*)0;
```
Comment (Rombur) [inline core/src/HIP/Kokkos_HIP_Space.cpp:452]:
```suggestion
  if (!arg_alloc_size) return nullptr;
```

File: core/src/HIP/Kokkos_HIP_Space.cpp [hunk 6]
```diff
@@ -445,6 +445,46 @@ void* SharedAllocationRecord<Kokkos::Experimental::HIPSpace, void>::
...
+
+  RecordBase::increment(r);
+
+  return r->data();
+}
+
+void SharedAllocationRecord<Kokkos::Experimental::HIPHostPinnedSpace,
+                            void>::deallocate_tracked(void* const
+                                                          arg_alloc_ptr) {
+  if (arg_alloc_ptr != 0) {
```
Comment (Rombur) [inline core/src/HIP/Kokkos_HIP_Space.cpp:465]:
```suggestion
  if (arg_alloc_ptr) {
```

File: core/src/HIP/Kokkos_HIP_Space.cpp [hunk 7]
```diff
@@ -477,6 +517,25 @@ SharedAllocationRecord<Kokkos::Experimental::HIPSpace, void>::get_record(
...
 
+SharedAllocationRecord<Kokkos::Experimental::HIPHostPinnedSpace, void>*
+SharedAllocationRecord<Kokkos::Experimental::HIPHostPinnedSpace,
+                       void>::get_record(void* alloc_ptr) {
+  using Header = SharedAllocationHeader;
+  using RecordHIP =
+      SharedAllocationRecord<Kokkos::Experimental::HIPHostPinnedSpace, void>;
+
+  Header* const h =
+      alloc_ptr ? reinterpret_cast<Header*>(alloc_ptr) - 1 : (Header*)0;
```
Comment (Rombur) [inline core/src/HIP/Kokkos_HIP_Space.cpp:528]:
```suggestion
      alloc_ptr ? reinterpret_cast<Header*>(alloc_ptr) - 1 : nullptr;
```

**Comment by dalg24** [comment]
OK to test

File: core/unit_test/TestTeamVector.hpp [hunk 8]
```diff
@@ -869,7 +869,7 @@ TEST(TEST_CATEGORY, team_vector) {
 #endif
 
 // FIXME_HIP
-#if !defined(KOKKOS_IMPL_CUDA_CLANG_WORKAROUND) && !defined(KOKKOS_ENABLE_HIP)
+#if !defined(KOKKOS_IMPL_CUDA_CLANG_WORKAROUND)
```
Comment (Rombur) [inline core/unit_test/TestTeamVector.hpp:872]: Remove the `FIXME_HIP`

**Comment by arghdos** [comment]
Thanks for the comments, will address tomorrow.  It seems like there are still some ghosts in the the Teams woodwork, getting a memory fault in `FixQEqReaxKokkosComputeHFunctor` in LAMMPS w/ a team-size of 32, and indefinite hanging for a team-size of 64 :)

**Comment by arghdos** [comment]
FYI that the following are failing in debug mode for me:

```GTEST_FILTER="-HIP.IncrTest_12a_ThreadScratch:hip.team_reduce:hip.team_broadcast_*```

It has to do w/ the `-DNDEBUG` interaction w/ hip_abort.  If we are using NDEBUG, the printf statement in `hip_abort` doesn't materialize, and the used register count is significantly reduced. w/out NDEBUG defined we are using too many registers for the hard-coded 1024 blocksize [workaround here](https://github.com/kokkos/kokkos/pull/3102/commits/e1262253a98351f7da1bfff9bd55c4f8b459a4f9#diff-07507380598cac4edcf9b8da4c73b53cR93).

Should these be guarded by `KOKKOS_ENABLE_DEBUG` or excluded in CMake somehow?

**Comment by arghdos** [comment]
Rebased onto develop, and addressed changes

File: core/src/HIP/Kokkos_HIP_BlockSize_Deduction.hpp [hunk 9]
```diff
@@ -156,23 +232,13 @@ int hip_get_opt_block_size(typename DriverType::functor_type const &f,
...
-  int const max_threads_per_block      = LaunchBounds::maxTperB == 0
-                                        ? max_threads_per_block_mi60
-                                        : LaunchBounds::maxTperB;
-  return std::min(attr.maxThreadsPerBlock, max_threads_per_block);
+int hip_get_opt_block_size(HIPInternal const *hip_instance,
+                           hipFuncAttributes const &attr, FunctorType const &f,
+                           size_t const vector_length, size_t const shmem_block,
+                           size_t const shmem_thread) {
+  return hip_internal_get_block_size<FunctorType, LaunchBounds>(
+      [](int x) { return true; }, hip_instance, attr, f, vector_length,
```
[the remark this answers is not in the document]
Comment (arghdos) [reply core/src/HIP/Kokkos_HIP_BlockSize_Deduction.hpp]: 542dd009c20044be0d9a4d44d048ba3b395c78f4

File: core/unit_test/TestTeamVector.hpp (hunk 8 above)
Comment (arghdos) [reply core/unit_test/TestTeamVector.hpp:870]: 774d6e75135122cab5f597855f7bd556fb77f36c

File: core/src/HIP/Kokkos_HIP_Space.cpp (hunk 7 above)
Comment (arghdos) [reply core/src/HIP/Kokkos_HIP_Space.cpp]: 774d6e75135122cab5f597855f7bd556fb77f36c

File: core/src/HIP/Kokkos_HIP_Space.cpp (hunk 6 above)
Comment (arghdos) [reply core/src/HIP/Kokkos_HIP_Space.cpp]: f88457a316ed5948e424e578423adfc7f984f7f5

File: core/src/HIP/Kokkos_HIP_Space.cpp (hunk 5 above)
Comment (arghdos) [reply core/src/HIP/Kokkos_HIP_Space.cpp]: f88457a316ed5948e424e578423adfc7f984f7f5

File: core/src/HIP/Kokkos_HIP_Parallel_Team.hpp (hunk 4 above)
Comment (arghdos) [reply core/src/HIP/Kokkos_HIP_Parallel_Team.hpp]: Restored in latest.

File: core/src/HIP/Kokkos_HIP_Parallel_Team.hpp (hunk 3 above)
Comment (arghdos) [reply core/src/HIP/Kokkos_HIP_Parallel_Team.hpp]: 7943a27511ba51cc14cd7a2c172613abb6041c43

File: core/src/HIP/Kokkos_HIP_Parallel_Range.hpp (hunk 2 above)
Comment (arghdos) [reply core/src/HIP/Kokkos_HIP_Parallel_Range.hpp]: 7943a27511ba51cc14cd7a2c172613abb6041c43

File: core/src/HIP/Kokkos_HIP_Abort.hpp (hunk 1 above)
Comment (arghdos) [reply core/src/HIP/Kokkos_HIP_Abort.hpp]: 4ada473cf081f87124619d11d2047fc942df40da

**Comment by Rombur** [comment]
> Should these be guarded by KOKKOS_ENABLE_DEBUG or excluded in CMake somehow?

You can guard them with `KOKKOS_ENABLE_DEBUG` and add `FIXME_HIP` with an explanation of the problem

File: core/src/HIP/Kokkos_HIP_Abort.hpp [hunk 10]
```diff
@@ -53,12 +53,13 @@
...
-__device__ inline void hip_abort(char const *msg) {
-  printf("%s", msg);
-  // FIXME_HIP both abort and the __assertfail system call are currently
-  // implemented with __builtin_trap which causes the program to exit abnormally
-  // without printing the error message.
-  // abort();
+__device__ __attribute__((noinline)) void hip_abort(char const *msg) {
+#ifndef NDEBUG
+  // disable printf on release builds, as it has a non-trivial performance
+  // impact
```
Comment (dalg24) [inline core/src/HIP/Kokkos_HIP_Abort.hpp:59]: Can you elaborate on how this would affect performance outside that function?

File: core/src/HIP/Kokkos_HIP_BlockSize_Deduction.hpp [hunk 11]
```diff
@@ -155,24 +231,16 @@ int hip_get_opt_block_size(typename DriverType::functor_type const &f,
                                             shmem_extra_thread);
 }
 
+template<class T> void ignore( const T& ) { }
```
Comment (dalg24) [inline core/src/HIP/Kokkos_HIP_BlockSize_Deduction.hpp:234]: Does the usual cast to void not work?

File: core/src/HIP/Kokkos_HIP_BlockSize_Deduction.hpp [hunk 12]
```diff
@@ -155,24 +231,16 @@ int hip_get_opt_block_size(typename DriverType::functor_type const &f,
...
-  int const max_threads_per_block      = LaunchBounds::maxTperB == 0
-                                        ? max_threads_per_block_mi60
-                                        : LaunchBounds::maxTperB;
-  return std::min(attr.maxThreadsPerBlock, max_threads_per_block);
+int hip_get_opt_block_size(HIPInternal const *hip_instance,
+                           hipFuncAttributes const &attr, FunctorType const &f,
+                           size_t const vector_length, size_t const shmem_block,
+                           size_t const shmem_thread) {
+  return hip_internal_get_block_size<FunctorType, LaunchBounds>(
+      [](int x) { ignore(x); return true; }, hip_instance, attr, f, vector_length,
```
Comment (masterleinad) [inline core/src/HIP/Kokkos_HIP_BlockSize_Deduction.hpp:242]:
```suggestion
      [](int) { return true; }, hip_instance, attr, f, vector_length,
```

**Comment by masterleinad** [comment]
Hmmm... I get multiple failing tests with this:
```
3: [  FAILED  ] hip.crs_copy_constructor
3: [  FAILED  ] hip.deep_copy_conversion
3: [  FAILED  ] hip.local_deepcopy_teampolicy_layoutleft
3: [  FAILED  ] hip.local_deepcopy_rangepolicy_layoutleft
3: [  FAILED  ] hip.local_deepcopy_teampolicy_layoutright
3: [  FAILED  ] hip.local_deepcopy_rangepolicy_layoutright
33: [  FAILED  ] hip.Random_XorShift64
33: [  FAILED  ] hip.Random_XorShift1024_0
33: [  FAILED  ] hip.SortUnsigned
32: [  FAILED  ] hip.offsetview_offsets_rank2
32: [  FAILED  ] hip.offsetview_offsets_rank3
32: [  FAILED  ] hip.staticcrsgraph
```

File: core/src/HIP/Kokkos_HIP_BlockSize_Deduction.hpp (hunk 11 above)
Comment (arghdos) [reply core/src/HIP/Kokkos_HIP_BlockSize_Deduction.hpp]: removed

File: core/src/HIP/Kokkos_HIP_Abort.hpp (hunk 10 above)
Comment (arghdos) [reply core/src/HIP/Kokkos_HIP_Abort.hpp:59]: Will discuss this morning

**Comment by arghdos** [comment]
@masterleinad -- I _think_ what's been going on is that I've been testing w/ -DNDEBUG defined, which eliminates the extra resources from the printf inside of abort().  If it's a register over-use problem, there's unfortunately not much we can do until [this issue](https://github.com/kokkos/kokkos/pull/3102/commits/e98b4bac1743234ecfdc87050a208c95ba43d213#diff-07507380598cac4edcf9b8da4c73b53cR93) gets fixed in the next release.  However.. those tests all pass for me in `-O3`, which is... odd:

```
GTEST_FILTER="*crs_copy_constructor:*deep_copy*:*deepcopy*:*Random*:*SortUnsigned:*offsetview:*staticcrsgraph" make test
...
100% tests passed, 0 tests failed out of 34

Total Test time (real) =  28.17 sec
```

File: core/src/HIP/Kokkos_HIP_BlockSize_Deduction.hpp (hunk 12 above)
Comment (arghdos) [reply core/src/HIP/Kokkos_HIP_BlockSize_Deduction.hpp]: 1deaf8d62a0e779d0175929d50d59bd2dcabcd09

File: core/src/HIP/Kokkos_HIP_Space.cpp [hunk 13]
```diff
@@ -514,6 +514,46 @@ void* SharedAllocationRecord<Kokkos::Experimental::HIPSpace, void>::
   return r_new->data();
 }
 
+void* SharedAllocationRecord<Kokkos::Experimental::HIPHostPinnedSpace, void>::
+    allocate_tracked(const Kokkos::Experimental::HIPHostPinnedSpace& arg_space,
+                     const std::string& arg_alloc_label,
+                     const size_t arg_alloc_size) {
+  if (!arg_alloc_size) return nullptr;
```
Comment (masterleinad) [inline core/src/HIP/Kokkos_HIP_Space.cpp:521]: Do you know where these overloads are used? Was there a test that would fail otherwise?

**Comment by masterleinad** [comment]
I would propose to split this pull request into smaller pieces so that we already merge working parts.

File: core/src/HIP/Kokkos_HIP_Space.cpp (hunk 13 above)
Comment (arghdos) [reply core/src/HIP/Kokkos_HIP_Space.cpp:521]: I believe they were required to compile LAMMPS

Comment (arghdos) [reply core/src/HIP/Kokkos_HIP_Space.cpp:521]: I believe they were needed for LAMMPS

**Comment by arghdos** [comment]
@masterleinad -- sounds good, definitely like 2-3 different things in here that can be separated.  Let's discuss this morning (along w/ the RDC / symbols etc.)

File: core/src/HIP/Kokkos_HIP_Space.cpp (hunk 13 above)
Comment (masterleinad) [reply core/src/HIP/Kokkos_HIP_Space.cpp:521]: I was asking because it seems that there is no test hitting this code path. We should add one.

**Comment by masterleinad** [comment]
>  I _think_ what's been going on is that I've been testing w/ -DNDEBUG defined, which eliminates the extra resources from the printf inside of abort()

Just incorporating the changes to `abort()` seems to work alright for me. I do see more test failures when we try to allow larger block sizes.

**Comment by arghdos** [comment]
Moved the HIPHostPinned and Abort stuff to https://github.com/kokkos/kokkos/pull/3137 and https://github.com/kokkos/kokkos/pull/3138 respectively.  Will leave this as a WIP until we get the occupancy / teams stuff working

**Comment by Rombur** [comment]
@arghdos for info @masterleinad has rebased this PR on master [here](https://github.com/kokkos/kokkos/pull/3165)

**Comment by masterleinad** [comment]
Superseded by https://github.com/kokkos/kokkos/pull/3165.
